### DISCUSSION TITLE
Afedri 4 channel coherent SDR - support?

### DISCUSSION THREAD

**Discussion by oh7eag** [AlexandreRouma/SDRPlusPlus#585] [Ideas]
Afedri is unique SDR in that it is available in a single box with 4 coherent and identical receivers. To an amateur, this opens wonderful possibilities that are only available in very expensive commercial or military equipment. I have been been experimenting for a long time now with an array of 4 identical HF antennas connected to Afedri. At the moment I am able e.g. to determine the azimuth and elevation angles of the incoming signal, next step is to add an adaptive algorithm for beam steering. However, I am not a s/w person, so the progress has been slow despite the fact that the source code (in C++/C) is available for the basic function of coherently combining the 4 channels.

Could this be something for SDR++?

**Comment by AlexandreRouma** [comment]
Hello,
Without having access to one of these SDRs I unfortunately can't add support for it.

As for direction finding it's not a planned freature

**Reply by oh7eag** [reply]
I am trying influence Alex to give you the h/w, after all it would be in his benefit. As regards to direction finding, I would not imagine it to be a feature of the s/w due to the hardware dependence. Instead, I would imagine that it would be attractive to build s/w that allows N number of incoming signals. One can do many useful things that simply are not possible with the other sdr software existing today. Unless, of course, your aim is simply to implement all the features of other software in as many hardware receivers and operating systems as possible?

**Comment by oh7eag** [comment]
Hi Alexandre,

I was in contact with Alex Trushkin about the possibility of giving out Afedri hardware for the purpose of the development of s/w for it. According to him, he has done this in several occasions, but his experiences have been negative, with no good results. He mentioned SDR Console as an example, it still does not have the support for dual channel Afedri.

Alex still seems to be willing to donate the h/w for s/w development if so requested, but the developers have to be responsive to his comments (the recent discussion in the Afedri .io group unfortunately went to a wrong direction) and there has to be feedback from the developers. He mentioned an example, where a programmer promised to develop s/w, got the Afedri h/w and was never heard of again.

If you are even a bit interested in this possibility, you should define first if you are interested in doing something that nobody else does, or if you want to capture the market and develop world's best software for everybody. For a single channel sdr there are so many alternatives available today that I personally don't see a reason for that kind of development. 

Coherent dual-channel receivers are becoming more popular and are available even from the main manufacturers, like Icom. Software development for such hardware would make sense. Linrad is a good example of what can be done in this area with open source s/w. Unfortunately its user interface is quite demanding for an average user. From my perspective the main limitation is that two channels are 'hardwired' in the s/w, and modifying it for 4 channels would be difficult, according to Leif, the developer of Linrad.

Afedri is the only commercially available low-cost sdr receiver that has 4 phase-coherent receiver channels in one box. From my professional background in microwave engineering I know the vast possibilities that such low-cost h/w can offer to a radio amateur. So after my retirement, I started a personal project for the development of a phase-steered 4 element antenna array for HF frequencies using Afedri as the receiver. 

First step was beam switching (4-square antenna): Using pre-calculated phase shifts between the 4 channels, the receive beam can be switched between pre-determined directions. Since I am using active antennas that are small wrt wavelength even at 30 MHz, practically the whole HF band is covered.

In the next step the receive beam can be continuously rotated (by changing the phase shifts) through 360 degrees in azimuth. By changing the phase shifts, a very deep null in the radiation pattern can also be generated, and it also can be continuously directed. This resembles a notch filter, but operates in the spatial domain.

I then added signal processing for the determination of the correlation matrix. From this I can e.g. determine the angles of arrival (both azimuth and elevation) of one, and in some cases of two incoming signals (in the same receive bandwidth). At the moment I am using MUSIC-algorithm, but am studying other options, too.

Next step is the development of adaptive algorithms for automatic maximisation of the wanted and elimination of interference signals, but this requires further studies of signal processing algorithms.

In this work the main difficulty (at least for a h/w engineer like myself) is the software: There is no sdr s/w available for a 4-channel receiver, except the one skeleton program provided by Alex, written in a mixture of C and C++ using Qt environment. What I have done is to modify Alex's program so that I can do the signal processing etc outside, presently in a Python program. So far this is ok, but there would clearly be need for proper programming.

There are a number of reasons why I have written the long story above:


1. To emphasize that Afedri 4-channel rx is unique, if you want to do something that others are not doing, go for it.

2. To give an impression of the possibilities that are available. The reason why not many people are working with it is the chicken-and-egg problem of h/w and s/w not being available.

3. To indicate the way on how the s/w for 4-ch rx could be organised: It is not realistic to develop a complete ready-made package due to the diversity of possible ways of using the rx with different applications and antenna configurations. More useful would be to develop s/w that provides the interface to the receiver h/w, does some basic operations like any other sdr s/w does for a single channel, and provides a clean programming interface for the users to develop their own subroutines for their specific signal processing applications.

Please, let me know what you think of this!

Best regards,

Tapani 
oh7eag





On Wednesday, 12 January 2022, 23:04:06 EET, AlexandreRouma ***@***.***> wrote: 







Hello,
Without having access to one of these SDRs I unfortunately can't add support for it.

As for direction finding it's not a planned freature

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by ecosierra51** [comment]
Hello!
I worked for many years with 2 and 4 channel Afedri, using LabVIEW. I am ready to help in this works.
Bests regards,
 ecosierra51

**Comment by oh7eag** [comment]
ecosierra51: Interesting, could you tell more?  Could you contact me directly: oh7eag@gmail.com
Tapani
